csrc/optimization/consecutive_cast.cpp

```diff
@@ -0,0 +1,198 @@
+//              `starting_anchor` and `lo_anchor` to both point at `val`.
+//
+//    1.4 At this point we look at `anchor_dtype` of `lo_anchor` and
+//    `output_dtype` of `expr->output(0)`:
+//
+//        a. if `anchor_dtype` is the same as `output_dtype`, we skip the last
+//        cast op and replace all its uses with `lo_anchor`;
+//
+//        b. if `anchor_dtype` is wider than `output_dtype`, all previous cast
+//        after `starting_anchor` is no-op, we re-wire `starting_anchor`
+//        directly to `expr`;
+//
+//        c. otherwise, we can't bypass `lo_anchor` cast, we rewire this
+//        section as `starting_anchor`->`lo_anchor`->`expr->output(0)`
+void castOptimizationPass(Fusion* fusion) {
+  // TODO: Traveral implies topological order on returned exprs, we can leverage
+  // that to improve the effieciency of the pass. In the case of a straight line
+  // casts, we are doing a lot of meaningless work here on mutating intermediate
+  // casts that would have been done again at the end of the chain.
+  // We should really use the reverse topological order and filters out exprs
+  // that has been rendered as dead code during the pass.
+  for (auto expr : fusion->exprs()) {
+    // skip current expr if it's not a foldable cast
+    if (!isCast(expr)) {
+      continue;
+    }
+    std::list<Val*> chain_cast_tvs;
+    auto prev_expr = expr->input(0)->definition();
+    while (prev_expr != nullptr && isCast(prev_expr)) {
+      auto intermediate_cast = prev_expr->output(0);
+      // 1.2 Note, if the output of prev_expr
+      //   is used by other operation(s); or
+      //   is a direct output from fusion
+      // we skip the casting chaining
+      if (intermediate_cast->isFusionOutput() ||
+          intermediate_cast->uses().size() > 1) {
+        break;
+      }
+
+      // in the loop, we just repetitively chaining consecutive casts.
+      chain_cast_tvs.push_front(intermediate_cast);
+      prev_expr = prev_expr->input(0)->definition();
+    }
+
+    // skip current expr if there's no chain_cast_tvs
+    if (chain_cast_tvs.empty()) {
+      continue;
+    }
+
+    // 1.3.1 Note, chain_cast_tvs has a straight-line use without branches
+    auto lo_anchor = chain_cast_tvs.front()->definition()->input(0);
+    auto anchor_dtype = lo_anchor->getDataType().value();
+    auto starting_anchor = lo_anchor;
+    for (auto val : chain_cast_tvs) {
+      auto val_dtype = val->getDataType().value();
+
+      // 1.3.2.a short-cut when we are not losing precision, either:
+      //   1. casting to the same type as the previously seen lowest precision;
+      //   or
+      //   2. casting to a wider type.
+      if (val_dtype == anchor_dtype || isWiderType(anchor_dtype, val_dtype)) {
+        continue;
+      }
+
+      // 1.3.2.c NOTE: To enter here, we have
+      //   !isWiderType(anchor_dtype, val_dtype) && isWiderType(val_dtype,
+      //   anchor_dtype)
```

Comment:
Are you saying the following?
```suggestion
      // 1.3.2.c NOTE: To enter here, we have
      //   !isWiderType(anchor_dtype, val_dtype) && !isWiderType(val_dtype,
      //   anchor_dtype)
```

Reply:
Good catch!